src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -801,7 +818,23 @@ CoordinatedRemoteTransactionsPrepare(void)
 			continue;
 		}
 
-		FinishRemoteTransactionPrepare(connection);
+		/*
+		 * Check if any DML or DDL is executed over the connection on any
+		 * placement/table. If yes, we finish preparing the transaction, otherwise
+		 * we skip prepare since the connection didn't perform any write (read-only)
+		 */
+		dlist_iter placementIter;
+		dlist_foreach(placementIter, &connection->referencedPlacements)
```

Comment:
And here instead of `dlist_foreach` the check could become:

```
if (connection->hasProcessedWrite) {
     FinishRemoteTransactionPrepare(connection);
}
```